website/src/react-luma/components/luma-example.tsx

```diff
@@ -63,52 +68,78 @@ export const LumaExample: FC<LumaExampleProps> = (props: LumaExampleProps) => {
-
-        let animationLoop: AnimationLoop | null = null;
-        animationLoop = makeAnimationLoop(props.template as unknown as typeof AnimationLoopTemplate, {
-          device,
-          autoResizeViewport: true,
-          autoResizeDrawingBuffer: true
-        });
 
-        // Start the actual example
-        animationLoop?.start();
-        setAnimationLoop(animationLoop);
-
-        // Ensure the example can find its images
-        // TODO - this only works for examples/tutorials
-        const RAW_GITHUB = 'https://raw.githubusercontent.com/visgl/luma.gl/master';
-        if (props.directory) {
-          setPathPrefix(`${RAW_GITHUB}/examples/${props.directory}/${props.id}/`);
-        } else {
-          setPathPrefix(`${RAW_GITHUB}/website/static/images/`);
-        }
-      }
+  useEffect(() => {
+    if (!canvas || usedCanvases.current.get(canvas)) return;
+
+    usedCanvases.current.set(canvas, true)
+
+    let animationLoop: AnimationLoop | null = null;
+    let device: Device | null = null;
+    const asyncCreateLoop = async () => {
+      if (animationLoopRef.current) return;
+
+      console.log(`starting example ${deviceType}`);
+      canvas.style.width = '100%';
+      canvas.style.height = '100%';
+      device = await luma.createDevice({type: deviceType, canvas, container: containerName});
 
-    } else {
+      animationLoop = makeAnimationLoop(props.template as unknown as typeof AnimationLoopTemplate, {
+        device,
+        autoResizeViewport: true,
+        autoResizeDrawingBuffer: true
+      });
 
-      if (animationLoop) {
-        console.error(`unmounting example ${props.id}`); // , ref.current);
-        animationLoop?.stop();
-        animationLoop?.destroy();
-        setAnimationLoop(null);
-        // animationLoop.device?.destroy();
+      // Start the actual example
+      animationLoop?.start();
+      animationLoopRef.current = animationLoop;
+
+      // Ensure the example can find its images
+      // TODO - this only works for examples/tutorials
+      const RAW_GITHUB = 'https://raw.githubusercontent.com/visgl/luma.gl/master';
+      if (props.directory) {
+        setPathPrefix(`${RAW_GITHUB}/examples/${props.directory}/${props.id}/`);
+      } else {
+        setPathPrefix(`${RAW_GITHUB}/website/static/images/`);
       }
-    }
-  }, []);
+    };
+
+
+    currentTask.current = Promise.resolve(currentTask.current).then(() => {
+      asyncCreateLoop().catch(error => {
+        console.log(`start ${deviceType} failed`, error);
+      });
+    });
+
+    return () => {
+      currentTask.current = Promise.resolve(currentTask.current)
+        .then(async () => {
+          console.log(`unmounting ${deviceType}`);
+          if (animationLoopRef.current) {
+            animationLoopRef.current.stop();
+            animationLoopRef.current.destroy();
+            animationLoopRef.current = null;
+          }
+
+          if (device) {
+            (device as any)?.destroy();
```

Comment:
Why is this cast needed? Seems like something we should fix.

Reply:
added destroy to Device